File: Data Structures & Algorithms/count-number-of-islands/submission-2.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty()) return 0;

        int ROWS = ssize(grid), COLS = ssize(grid[0]);
        set<pair<int, int>> visited_land;
        int island_count = 0;

        function<void(int r, int c)> dfs = [&](int r, int c) -> void {
            if (min(r, c) < 0 || 
                r == ROWS || 
                c == COLS || 
                visited_land.contains({r, c}) || 
                grid[r][c] == '0') return;

            visited_land.insert({r, c});
            dfs(r - 1, c);
            dfs(r, c + 1);
            dfs(r + 1, c);
            dfs(r, c - 1);
        };

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == '1' && !visited_land.contains({r, c})) {
                    dfs(r, c);
                    island_count++;
                }
            }
        }

        return island_count;
    }
};

File: Data Structures & Algorithms/max-area-of-island/submission-2.cs
public class Solution {
    public int MaxAreaOfIsland(int[][] grid) {
        int ROWS = grid.Length, COLS = grid[0].Length;
        HashSet<(int r, int c)> visitedLand = new HashSet<(int r, int c)>();
        int maxArea = 0;

        int Dfs(int r, int c) {
            if (Math.Min(r, c) < 0 || 
                r == ROWS || 
                c == COLS || 
                grid[r][c] == 0 || 
                visitedLand.Contains((r, c))) return 0;

            visitedLand.Add((r, c));
            int currArea = 1;
            currArea += Dfs(r - 1, c);
            currArea += Dfs(r, c + 1);
            currArea += Dfs(r + 1, c);
            currArea += Dfs(r, c - 1);

            return currArea;
        }

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == 1 && !visitedLand.Contains((r, c))) {
                    maxArea = Math.Max(maxArea, Dfs(r, c));
                }
            }
        }
        return maxArea;
    }
}
